### ISSUE TITLE
Sandboxing with luerl?

### ISSUE THREAD

**Issue by tisba** [rvirding/luerl#53]
Disclaimer: I'm not terribly familiar with lua :)

Hi,

I'm currently evaluating how to run customer provided code/functions within a sandbox. I was also taking a look at JavaScript (via https://github.com/basho/erlang_js), but I somehow like lua better.

I was wondering, if there is any way to "seal" off parts of luas standard library, like access to filesystem etc.

**Comment by rvirding** [comment]
I too prefer Lua to JS, which is why I did luerl instead jserl.

With Lua sandboxing is relatively simple because of the way Lua finds things in the global environment. Everything is in a tree of tables originating from one top table. However when you compile a module all the global variables are modified to go through an external local variable _ENV, so foo becomes _ENV.foo. The _ENV is given a value when the module is loaded so at load time you have full control of which global environment a module has. For example you can create one without any functions or modules which allow you to compile/load code.

This is done at a per module basis so you can see access everything but when you want to load in user code which should be sandboxed you can create an environment containing just what you want them to see and this become *their* global environment. Check out section "2.2 – Environments and the Global Environment" in the Lua manual or ask the Lua mailing list for more information.

Luerl doesn't fully implement this yet as no-one has asked for it. When a use actually wants/needs it then I can quite easily fix it.

I hope this helps.

Robert

**Comment by tisba** [comment]
Thanks!

I think I need to get a little bit more familiar with lua and luerl to figure out how this could work (e.g. I didn't quite understand what luerl is missing currently to do that).

I'm not sure what use cases for luerl are out there, but I think sandboxing is not that uninteresting. So maybe it's worth having an outline or example for that? I'm happy to help with that if I can.

**Comment by t-mw** [comment]
I'm quite interested in sandboxing untrusted code with luerl, and have tried the following (from http://lua-users.org/wiki/SandBoxes):

```lua
-- run code under environment [Lua 5.2]
local function run(untrusted_code)
    local untrusted_function, message = load(untrusted_code, untrusted_code, 't', env)
    if not untrusted_function then return nil, message end
    return pcall(untrusted_function)
end
```

However the `env` argument to `load` doesn't seem to replace the upvalues in the untrusted function, and so the global scope is still accessible from the function. From your previous reply, I'm assuming that's because of the way `_ENV` is handled in luerl, so I'd be happy to help by either writing tests or trying to implement it myself given some pointers on where to look.

**Comment by lastcanal** [comment]
It's not necessary to protect the global table with luerl unless you plan on reusing the state which has run your sandboxed script. Give each sandboxed script a clean copy of your luerl state for each run. The state can be kept in an ets table with `read_concurrency` for each script runner grab before executing.

One way to build a sandboxed state is by removing the unsafe global tables and functions that are listed in the lua sandboxes wiki page:

```erlang
init_sandbox() ->
  Globals = [io, file, os, package, load, loadfile, require, dofile],
  init_sandbox(Globals, luerl:init()).

init_sandbox([], St) -> St;
init_sandbox([Global|T], St) ->
  init_sandbox(T, luerl:set_table(['_G', Global], sandboxed, St)).
```

**Comment by rvirding** [comment]
The handling of _ENV would be pretty deep inside the Luerl implementation so is it not trivial to add it.

It is actually not needed either. The Luerl implementation only allows access to the Erlang "outside" through modules written in Erlang. There are no modules in the current implementation which allow general access to the "outside", to make general Erlang calls, only to specific things. It is impossible load in modules written in Erlang into the Luerl system from inside Luerl, you can only load Lua modules.

Each Luerl state is a normal Erlang data structure which can be created for a Lua invocation and then thrown away later. It cannot be shared between Lua invocations, each one is independent. If you are running multiple Luerl processes then by necessity each process will be an independent Lua machine which does not share any state with other Luerl process/invocation. This is guaranteed by how Luerl is implemented and how Erlang works. Erlang does not do global or sharing.

The only thing you may want to do is remove some tables which do access the outside world, for example io or file. Loading Lua modules is not a problem.

The only way to load potentially dangerous code is from Erlang outside of Lua code in which case sandboxing is basically impossible anyway.

t-mw, what type of things are you interested in doing?

**Comment by rvirding** [comment]
Sorry, that became a bit garbled. To make it clearer (I hope):

- To open up the Luerl system you need modules written in Erlang, modules in Lua cannot get out by themselves.
- From inside Luerl you can only load modules written in Lua which cannot by themselves get out.
- From outside Luerl you can load modules written in Erlang into your environment.
- You cannot share Luerl state between invocations of Luerl, each runs in its own private environment.
- None of the standard Leuel modules allow you free access to the Erlang outside, they are all specific.
- Very few of the standard modules pose any problems, probably only io and file. Os is safe.

That is about all. So the way Luerl is implemented means that it is sandboxed.

**Comment by t-mw** [comment]
Thanks both for the detailed replies.

My use case is allowing users to modify data in a Lua table with their own code, but with access to keys being restricted: some keys are read-only, while others sanitise their values before being set. My first attempt at achieving that in a native VM involves using metatables to override the indexing metamethods. I then make those restricted versions of the tables and certain other functions in the module available to the user code through the `env` parameter in `load`. The same tables, having been interacted with by the previous user, are then provided to the next user to access, so state is persistent across script invocations.

In short my use case involves sandboxing untrusted code from the global Lua scope, rather than just the OS. Given that my Lua script state is persistent, using Lua through a C node might be a better fit for my purposes. Although I should be able to achieve the same effect in luerl by overriding `_G` instead of `_ENV`, perhaps by iterating through every available key in `_G` in Erlang and clearing it from the table.

**Comment by Licenser** [comment]
I'm really interested in this too, has anyone had any success with this where I can be a lazy person and take a peak at how others do it? </lazy> would be happy to help write up a wiki entry (all typos and spelling mistakes are free additions) with having something to start off at.

**Comment by Licenser** [comment]
So reading some code, and understanding a bit more about luerl this seems to be a good approach to create a 'safe' sandbox, I could make a PR to ad the init_sandbox to the luerl module if there is interest?

```erlang
init_sandbox() ->
  L0 = luerl:init(),
  L1 = luerl:set_table([io], sandbox, L0),
  luerl:set_table([file], sandbox, L0).
```

**Comment by rvirding** [comment]
This is pretty specific and could easily change over time. I think a better solution would be have a section in the docs describing how to interface luerl to the Erlang outside. Such a section would also discuss sandboxing which would be to null or remove some tables. There might also be some discussion about monitoring the luerl processes to make check that they don't allocate too many resources, for example memory.

**Comment by jchassoul** [comment]
I think that what you propose around documentation is the better solution, but the documentation in general around luerl needs some improvements.

Currently it is complete and full of technical information that matter but is confusing for new comers or even the common lua dev in general I mean this recalling the coroutine issues and I agree that implementing coroutines in luerl not make sense.

What needs to be very clear is those type of knowledge I know this in general is common sense from the eyes of a Erlang developer but common devs don't know much about the greatness of the beam and its similarities with a operating system.

Call it just a whisper in my ghost but I believe improvements on that regard will bring more users to luerl, but it's just matter of time really, this project is awesome and deserves all the future love.

**Comment by rvirding** [comment]
I quite agree about the documentation, there is in fact very little. My only excuse is lack of time.

If someone is interested in attacking this problem I would be very grateful. Even if it is just a start with some notes we could extend to proper documentation in the future.

**Comment by Licenser** [comment]
I've tarted a little wiki page. I'm not the best writer but I wrote down a few things that I learned over the past days and would have found helpful to read :)

https://github.com/rvirding/luerl/wiki/Interfacing-with-Erlang

**Comment by rvirding** [comment]
Thank you. I didn't know I had made it so others could add pages. I do the right thing without even knowing it. 😄 Please feel free to continue. My only request is that you make sure you sign it so the credit goes where it is due.

**Comment by rvirding** [comment]
Good.
